Initialize the English lecture 4 (#3)
moonbitlang/moonbit-course: course4/lecture4_en.md

> # Modern Programming Concepts: Tuples, Structures, and Enumerated Types

Reviewer: ```suggestion
# Modern Programming Ideology: Tuples, Structures, and Enumerated Types
```
As suggested by @peter-jerry-ye , the course title should be "Modern Programming Ideology".

Hello everyone, welcome to the open course on modern programming concepts brought to you by the IDEA's Digital Infrastructure Innovation. Today's topic is tuples, structures, and enumerated types. In a nutshell, we will be exploring how to customize data structures in MoonBit.

> # Review: Tuples

Reviewer 2: Maybe use a second level heading here and for everything else? There should be only one H1 in one document.

A MoonBit tuple is a composite of data which have different types with a fixed length. In comparison, lists are collections of data which have the same type with arbitrary lengths. For example, the length of the list below is not fixed, but the values stored must all be of the character type. In the previous lesson, we didn't discuss why `Cons` is named `Cons`. It is an abbreviation for `Construct`.

```moonbit
Cons('H', Cons('i', Cons('!', Nil)))
```

The definition of tuples is represented by expressions enclosed in parentheses and separated by commas. The types are also specified using the same syntax, as in the definition of personal identity information here: `("Bob", 2023, 10, 24): (String, Int, Int, Int)`.

Tuple members are accessed by indexes, starting from 0 on the left. For example, `(2023, 10, 24).0 == 2023`.

# Cartesian Product

You may be familiar with the concept of the Cartesian product. The Cartesian product of two sets is a set where all elements are ordered pairs formed by the elements of the original two sets. For example, the Cartesian product of the set of card suits and the numbers 1 to 52 forms all possible ordered pairs of suits and numbers. 

> { ♥️ ♦️ ♠️ ♣️ } $\times \{ n \in \mathbb{N} | 1 \leq n \leq 52 \}$ 

Reviewer 2: ```suggestion
${ ♥️ ♦️ ♠️ ♣️ } $\times \{ n \in \mathbb{N} | 1 \leq n \leq 52 \}$ 
```

Author: { ♥️ ♦️ ♠️ ♣️ } $\times \lbrace n \in \mathbb{N} | 1 \leq n \leq 52 \rbrace$ 

`\{` and `\}` seem not to be rendered on Github. So I replace them with `\lbrace` and `\rbrace`

> Tuples, on the other hand, go beyond the Cartesian product of two sets; they represent the Cartesian product of multiple sets, making them more accurately termed as ordered sets. Consequently, tuples are also known as product types. You might wonder if there are sum types alongside product types. We will introduce sum types and explore the concepts of **zero** and **one** later.

Reviewer 2: I didn't know ordered sets means cartesian product of multiple sets. May I have a resource please?

Author: My mistake here. The ordered set is not the idea we need here. Maybe the following is more appropriate: 

Beyond the Cartesian product of two sets tuples represent the Cartesian product of multiple sets, named n-tuple. Consequently, tuples are also known as product types. You might wonder if there are sum types alongside product types. We will introduce sum types and explore the concepts of **zero** and **one** later.

Reviewer 2: I think I now understand what you meant by ordered sets. n-tuple may be better.

# Structures

The problem is that it is hard to understand the data represented by tuples. For example, `(String, Int)` – does it represent a person's name and age, or a person's name and phone number, or perhaps address and email?

Structures allow us to assign **names** to the data, both to the entire type and to each field individually. For instance:

- `struct PersonalInfo { name: String; age: Int }`
- `struct ContactInfo { name: String; telephone: Int }`
- `struct AddressInfo { address: String; postal: Int }`

We can clearly understand the information about the data and the meaning of each corresponding field by names.

The syntax of defining a structure is `struct <struct_name> { <field_name>: <type> ; ... }`. For example: `struct PersonalInfo { name: String; age: Int }`. As we have mentioned, the semicolons can be omitted when the fields are on separate lines.

> Definitions of the values of a structure are enclosed in braces, with each field assigned a value: `<field_name>: <value>`; each field assignment is followed by a comma. For instance: `let info: PersonalInfo = { name: "Moonbit", age: 1, }`. Note the comma after each value. It can be omitted if the last comma is directly followed by the closing brace without a newline. The order of fields does not matter, for example, `{ age: 1, name: "Moonbit" }`. If two structures happen to have exactly the same field names and types, they cannot be distinguished solely by the defined values. In such cases, we can differentiate them by adding a type declaration after the structure's value, such as `: A`.

Reviewer 2: Maybe a more complete example than just `: A` would be better?

Author: How about:

If two structures have the exact same fields with the same types, it's challenging to tell them apart just by their field values. To resolve this, we add type annotations like `let jack: PersonInfo = { ... }` to specify the type.

Reviewer 2: I was thinking about an example where you may have

`let jack = ({ ... } : PersonalInfo)` just to demonstrate how to specify the type of a value (rather than the type of an identifier).

```moonbit
fn accept(tuple: (Int, String)) -> Bool {
  true
}
let accepted: Bool = accept((1, "Yes"))
```

On the other hand, structures are *nominal*, meaning compatibility is based on the type name, and the internal order can be rearranged. In the first example, even though the structures are identical, the function cannot accept the structure because the type names are different. In the second example, the function can accept it because the types are the same even if the order of the fields is different.

```moonbit
struct A { val : Int ; other: Int }
struct B { val : Int ; other: Int }
fn accept(a: A) -> Bool {
  true
}
let not_accepted: Bool = accept(({ val : 1, other : 2 }: B)) // DO NOT COMPILE
let accepted: Bool = accept(({other: 2, val: 1}: A))
```

# Pattern Matching

Pattern matching is another way to access tuples and structures.

```moonbit
fn head_opt(list: List[Int]) -> Option[Int] {
  match list {
    Nil => None
    Cons(head, tail) => Some(head)
  }
}

fn get_or_else(option_int: Option[Int], default: Int) -> Int {
  match option_int {
    None => default
    Some(value) => value
  }
}
```

> We have previously used pattern matching to inspect the structure of lists and Options. For instance, using `Nil` and `Cons` to match lists; `None` and `Some` to match Options. In fact, pattern matching can match values (logical values, numbers, characters, strings) as well as constructors. 

Reviewer 2: ```suggestion
We have previously used pattern matching to inspect the structure of `List` and `Option`. For instance, using `Nil` and `Cons` to match lists; `None` and `Some` to match Options. In fact, pattern matching can match values (booleans, numbers, characters, strings) as well as constructors. 
```

```moonbit
fn is_zero(i: Int) -> Bool {
  match i {
    0 => true
    1 | 2 | 3 => false
    _ => false
  }
}
```

In the examples above, we matched numbers. Here we use the pipe symbol (the 'or' pattern) to simultaneously match multiple possible values. The underscore (_) is the wildcard to match all remaining cases. We can nest patterns in constructors, or bind corresponding structures with identifiers.

```moonbit
fn contains_zero(l: List[Int]) -> Bool {
  match l {
    Nil => false
    Cons(0, _) => true
    Cons(_, tl) => contains_zero(tl)
  }
}
```

> In this example, the first `Cons` uses the number 0 to match lists starting with 0, and the second `Cons` uses a wildcard and an identifier `tl` to match remaining lists while binding the sublist to the identifier `tl` for further processing. The value of the current list is discarded by the wildcard.

Reviewer 2: I think it is a bit imprecise to say 'the value of the current list'. Maybe the head of the current list or the first element of the current list?
Also, `Cons` does not match 'remaining lists' -> it just matches the whole list

Author: Yes, `Cons` matches the whole list and I know that. My narrative caused ambiguity here. I meant the second `Cons` will match the rest of the lists left by the first `Cons` case. How about this revision: 

In this example, the first `Cons` match lists starting with 0 with the literal number 0. The second `Cons` uses a wildcard to match other lists, while binding the sublist to the identifier `tl` for further processing. The head of the current list is discarded by the wildcard.

Reviewer 2: Yes, this is better. Also I think you can just write it out as 

> the branch `Cons(0, _)`
> the branch `Cons(_, tl)`

```moonbit
fn zip(l1: List[Int], l2: List[Char]) -> List[(Int, Char)] {
  match (l1, l2) {
    (Cons(hd, tl), Cons(hd2, tl2)) => Cons((hd, hd2), zip(tl, tl2))
    _ => Nil
  }
}
```

We define our function with pattern matching. Here, we match a pairs by constructing a tuple and then match the nested tuple pattern, effectively matching both lists simultaneously. If either of the input lists is empty, the result is an empty list. When both lists are non-empty, we get a non-empty result. The first item of the result is a tuple of the two values we take from the inputs, followed by the zipped result of the sublists of both lists. Note that the order of pattern matching is top-down. (If a wildcard is placed at the top, the subsequent patterns will never be matched, and the code will never run. The good news is that MoonBit can detect this and provide warnings. These warnings are advisory and won't prevent compilation, so it's crucial to pay attention to the issues panel in your IDE.)

> Lastly, pattern matching is not limited to `match`; it can also be used in data binding. In local definitions, we can use pattern matching expressions to bind corresponding substructures to identifiers. It's essential to note that if value comparison occurs at this point and the match fails, the program will encounter a runtime error, leading to program termination.

Reviewer 2: Maybe an example here? Also what does 'if value comparison occurs' mean?

Author: The example of *let binding pattern*? (The zip function definition is above)

'if value comparison occurs' is directly translated from Chinese. I think what we want to express here is the difference between 'refutable/irrefutable' patterns. But I am not sure if they are standard terminologies in MoonBit.

Reviewer 2: > The example of let binding pattern? (The zip function definition is above)

Yes

> 'if value comparison occurs' is directly translated from Chinese.

Ok, I see what you mean. Maybe we can just omit it and say 'if the match fails'

# Enumerated Types

Now, let's delve into the enumerated types.

Think about this, how should we represent the union of several possibilities? For example, how do we define a type that represents the set of days from Monday to Sunday? How about defining a type for the outcomes of a coin toss – heads or tails? What about a type to represent the results of integer arithmetic operations, such as a successful result, overflow, or division by zero?

The answer is enumerated types. Enumerated types allow us to define data structures that represent different cases. For example, we define a collection of seven days of the week here and the outcomes of a coin toss.

```moonbit
enum DaysOfWeek {
  Monday; Tuesday; Wednesday; Thursday; Friday; Saturday; Sunday
}

enum Coin {
  Head
  Tail
}
```

The construction of an enumerated type is as follows:

```moonbit
enum <type_name> { <variant>; }
```

> Here, each possible variant is a constructor. For instance, `let monday = Monday`, where `Monday` defines the day of the week as Monday. Different enumerated types may cause conflicts because they might use the same names for some cases. In such cases, we distinguish them by adding `<type>::` in front of the constructor.

Reviewer 2: ```suggestion
Here, each possible variant is a constructor. For instance, `let monday = Monday`, where `Monday` defines the day of the week as Monday. Different enumerated types may cause conflicts because they might use the same names for some cases. In such cases, we distinguish them by adding `<type>::` in front of the constructor, such as `DaysOfWeek::Monday`.
```

Now we need to ask, why do we need enumerated types? Why not just use numbers from one to seven to represent Monday to Sunday? Let's compare the following two functions. 

```moonbit
fn tomorrow(today: Int) -> Int
fn tomorrow(today: DaysOfWeek) -> DaysOfWeek
let tuesday = 1 * 2 // Is this Tuesday?
```

> The most significant difference is that functions defined with enumerated types are total functions, while those defined with integers are partial functions. This increases the possibility of users providing incorrect inputs – they might pass -1 or 8, and we have no way to prevent this through the compiler. Another consideration is, what does adding one to a day of the week mean? What is the meaning of multiplying the day of the week by a number? Why is Monday multiplied by two equal to Tuesday? Why is Tuesday divided by two equal to Monday? Enumerated types can distinguish themselves from existing types and better abstraction.

Reviewer 2: ```suggestion
The most significant difference is that functions defined with enumerated types are total functions, while those defined with integers are partial functions. This increases the possibility of users providing incorrect inputs – they might pass -1 or 8, and we have no way to prevent this through the compiler. Another consideration is, what does adding one to a day of the week mean? What is the meaning of multiplying the day of the week by a number? Why is Monday multiplied by two equal to Tuesday? Why is Tuesday divided by two equal to Monday? Enumerated types can distinguish themselves from existing types and abstract better.
```

Here are a few examples. 

```moonbit
enum Coins { Head; Tail }
```

$\texttt{Coins} = 1 + 1 = 2$

```moonbit
enum DaysOfWeek { Monday; Tuesday; ...; }
```

$\texttt{DaysOfWeek} = 1 + 1 + 1 + 1 + 1 + 1 + 1 = 7$

The data type for the coin toss can be considered as 1 + 1, as each case, `Head` and `Tail`, actually represents a set with only one value. Therefore, each case is isomorphic to the Unit type. When combined by the sum type, the `Coin` type becomes 1 + 1 = 2, representing a set with two values, which is isomorphic to any other type with two values. Similarly, `DaysOfWeek` represents a set of seven values, isomorphic to any other type with seven values. 

A more interesting example is `List`, using `List[Int]` as an example. 

$$
\begin{array}{rl}

> \texttt{enum} \;\; \texttt{List} & = \texttt{Nil} + \texttt{Int} \times \texttt{List} \\  

Reviewer 2: I think the preview of github does not support `\;`. Is there any better alternative?

Author: Just `\ ` is OK. 

$$
\begin{array}{rl}
\texttt{enum} \ \ \texttt{List} & = \texttt{Nil} + \texttt{Int} \times \texttt{List} \\  
& = \texttt{1} + \texttt{Int} \times \texttt{List} \\
& = \texttt{1} + \texttt{Int} \times (\texttt{1} + \texttt{Int} \times \texttt{List} ) \\
& = \texttt{1} + \texttt{Int} \times \texttt{1} + \texttt{Int} \times \texttt{Int} \times \texttt{List} \\
& = \texttt{1} + \texttt{Int} + \texttt{Int} \times \texttt{Int} + \texttt{Int} * \texttt{Int} \times \texttt{Int} + ...
\end{array}
$$

# Summary

In this section, we explored various custom data types in MoonBit, including:

- **Tuples:** Fixed-length combinations of different data types.
- **Structures:** Tuples with names to fields for better understanding.
- **Enumerated Types:** Types that represent a distinct set of values, often used to model different cases or options.
  
We also touched upon the concept of algebraic data types, which encompass tuples, structures, and enumerated types, and discussed some basic properties resembling those found in algebra.

For further exploration, we recommend reading Chapter 6 of [Category Theory for Programmers](https://bartoszmilewski.com/2015/01/13/simple-algebraic-data-types/).

> You can also experiment with MoonBit: [try.moonbitlang.com/#44a4eb28](https://try.moonbitlang.com/#44a4eb28).

Reviewer 2: ```suggestion
You can also experiment with MoonBit: [try.moonbitlang.com](https://try.moonbitlang.com/course/lec4/data_structures.mbt).
```
